File: dart/lib/src/provider_service.dart
import 'dart:convert';

import 'package:trinsic_dart/src/proto/sdk/options/v1/options.pb.dart';
import 'package:trinsic_dart/src/proto/services/provider/v1/provider.pbgrpc.dart';
import 'package:trinsic_dart/src/service_base.dart';

class ProviderService extends ServiceBase {
  late ProviderClient client;

  ProviderService(ServiceOptions? serverOptions) : super(serverOptions) {
    client = ProviderClient(super.channel);
  }

  Future<CreateEcosystemResponse> createEcosystem(
      {CreateEcosystemRequest? request}) async {
    request = request ?? CreateEcosystemRequest();
    var authenticate =
        (request.name.isNotEmpty || request.details.email.isNotEmpty);
    var metadata =
        authenticate ? buildMetadata(request: request) : buildMetadata();
    var response = await client.createEcosystem(request, options: metadata);
    serviceOptions.authToken =
        Base64Encoder.urlSafe().convert(response.profile.writeToBuffer());
    return response;
  }

// BEGIN Code generated by protoc-gen-trinsic. DO NOT EDIT.
// target: ..\sdk\dart\lib\src\provider_service.dart

  Future<UpdateEcosystemResponse> updateEcosystem(
      UpdateEcosystemRequest request) async {
    ///  Update an existing ecosystem
    return await client.updateEcosystem(request,
        options: buildMetadata(request: request));
  }

  Future<GrantAuthorizationResponse> grantAuthorization(
      GrantAuthorizationRequest request) async {
    ///  Grant user authorization to ecosystem resources
    return await client.grantAuthorization(request,
        options: buildMetadata(request: request));
  }

  Future<RevokeAuthorizationResponse> revokeAuthorization(
      RevokeAuthorizationRequest request) async {
    ///  Revoke user authorization to ecosystem resources
    return await client.revokeAuthorization(request,
        options: buildMetadata(request: request));
  }

  Future<GetAuthorizationsResponse> getAuthorizations(
      GetAuthorizationsRequest request) async {
    ///  Retreive the list of permissions for this particular account/ecosystem
    return await client.getAuthorizations(request,
        options: buildMetadata(request: request));
  }

  Future<AddWebhookResponse> addWebhook(AddWebhookRequest request) async {
    ///  Add a webhook endpoint to the ecosystem
    return await client.addWebhook(request,
        options: buildMetadata(request: request));
  }

  Future<DeleteWebhookResponse> deleteWebhook(
      DeleteWebhookRequest request) async {
    ///  Delete a webhook endpoint from the ecosystem
    return await client.deleteWebhook(request,
        options: buildMetadata(request: request));
  }

  Future<EcosystemInfoResponse> ecosystemInfo(
      EcosystemInfoRequest request) async {
    ///  Get ecosystem information
    return await client.ecosystemInfo(request,
        options: buildMetadata(request: request));
  }

  Future<GenerateTokenResponse> generateToken(
      GenerateTokenRequest request) async {
    ///  Generates an unprotected authentication token that can be used to
    /// configure server side applications
    return await client.generateToken(request,
        options: buildMetadata(request: request));
  }

  Future<InviteResponse> invite(InviteRequest request) async {
    ///  Invite a user to the ecosystem
    return await client.invite(request,
        options: buildMetadata(request: request));
  }

  Future<InvitationStatusResponse> invitationStatus(
      InvitationStatusRequest request) async {
    ///  Check the status of an invitation
    return await client.invitationStatus(request,
        options: buildMetadata(request: request));
  }

  Future<GetOberonKeyResponse> getOberonKey(GetOberonKeyRequest request) async {
    ///  Returns the public key being used to create/verify oberon tokens
    return await client.getOberonKey(request, options: buildMetadata());
  }

  Future<GetEventTokenResponse> getEventToken(
      GetEventTokenRequest request) async {
    ///  Generate a signed token (JWT) that can be used to connect to the message bus
    return await client.getEventToken(request,
        options: buildMetadata(request: request));
  }
// END Code generated by protoc-gen-trinsic. DO NOT EDIT.
}

File: web/src/ProviderService.ts
import ServiceBase from "./ServiceBase";
import * as proto from "./proto";

import type { Client as BrowserClient } from "nice-grpc-web";
import { Base64 } from "js-base64";

export class ProviderService extends ServiceBase {
    client: BrowserClient<typeof proto.ProviderDefinition>;

    constructor(options?: proto.TrinsicOptions) {
        super(options);

        this.client = this.createClient(proto.ProviderDefinition);
    }

    public async createEcosystem(
        request: proto.CreateEcosystemRequest,
    ): Promise<proto.CreateEcosystemResponse> {
        const response =
            request.name?.trim() || request.details?.email?.trim()
                ? await this.client.createEcosystem(request, {
                      metadata: await this.buildMetadata(
                          proto.CreateEcosystemRequest.encode(request).finish(),
                      ),
                  })
                : await this.client.createEcosystem(request);

        const authToken = Base64.fromUint8Array(
            proto.AccountProfile.encode(response.profile!).finish(),
            true,
        );
        this.options.authToken = authToken;
        return response;
    }
// BEGIN Code generated by protoc-gen-trinsic. DO NOT EDIT.
// target: /home/runner/work/sdk/sdk/web/src/ProviderService.ts

  /** Returns the public key being used to create/verify oberon tokens */
  public async getOberonKey(): Promise<proto.GetOberonKeyResponse> {
    let request = proto.GetOberonKeyRequest.fromPartial({});
    return this.client.getOberonKey(request, {
      metadata: await this.buildMetadata()
    });
  }
  /** Upgrade a wallet's DID from `did:key` to another method */
  public async upgradeDID(request: proto.UpgradeDidRequest): Promise<proto.UpgradeDidResponse> {
    
    return this.client.upgradeDID(request, {
      metadata: await this.buildMetadata(proto.UpgradeDidRequest.encode(request).finish())
    });
  }
  /** Search for issuers/providers/verifiers in the current ecosystem */
  public async searchWalletConfigurations(request: proto.SearchWalletConfigurationsRequest): Promise<proto.SearchWalletConfigurationResponse> {
    
    return this.client.searchWalletConfigurations(request, {
      metadata: await this.buildMetadata(proto.SearchWalletConfigurationsRequest.encode(request).finish())
    });
  }
// END Code generated by protoc-gen-trinsic. DO NOT EDIT.
}
